src/cascadia/TerminalApp/Tab.cpp

```diff
@@ -513,15 +519,164 @@ namespace winrt::TerminalApp::implementation
             }
         });
 
+        Controls::MenuFlyoutItem renameTabMenuItem;
+        {
+            // "Rename Tab"
+            Controls::FontIcon renameTabSymbol;
+            renameTabSymbol.FontFamily(Media::FontFamily{ L"Segoe MDL2 Assets" });
+            renameTabSymbol.Glyph(L"\xE932"); // Label
+
+            renameTabMenuItem.Click([weakThis](auto&&, auto&&) {
+                if (auto tab{ weakThis.get() })
+                {
+                    tab->_inRename = true;
+                    tab->_UpdateTabHeader();
+                }
+            });
+            renameTabMenuItem.Text(RS_(L"RenameTabText"));
+            renameTabMenuItem.Icon(renameTabSymbol);
+        }
+
         // Build the menu
         Controls::MenuFlyout newTabFlyout;
         Controls::MenuFlyoutSeparator menuSeparator;
         newTabFlyout.Items().Append(chooseColorMenuItem);
+        newTabFlyout.Items().Append(renameTabMenuItem);
         newTabFlyout.Items().Append(menuSeparator);
         newTabFlyout.Items().Append(closeTabMenuItem);
         _tabViewItem.ContextFlyout(newTabFlyout);
     }
 
+    // Method Description:
+    // - This will update the contents of our TabViewItem for our current state.
+    //   - If we're not in a rename, we'll set the Header of the TabViewItem to
+    //     simply our current tab text (either the runtime tab text or the
+    //     active terminal's text).
+    //   - If we're in a rename, then we'll set the Header to a TextBox with the
+    //     current tab text. The user can then use that TextBox to set a string
+    //     to use as an override for the tab's text.
+    // Arguments:
+    // - <none>
+    // Return Value:
+    // - <none>
+    void Tab::_UpdateTabHeader()
+    {
+        winrt::hstring tabText{ GetActiveTitle() };
+
+        if (!_inRename)
+        {
+            // If we're not currently in the process of renaming the tab, then just set the tab's text to whatever our active title is.
+            _tabViewItem.Header(winrt::box_value(tabText));
+        }
+        else
+        {
+            _ConstructTabRenameBox(tabText);
+        }
+    }
+
+    // Method Description:
+    // - Create a new TextBox to use as the control for renaming the tab text.
+    //   If the text box is already created, then this will do nothing, and
+    //   leave the current box unmodified.
+    // Arguments:
+    // - tabText: This should be the text to initialize the rename text box with.
+    // Return Value:
+    // - <none>
+    void Tab::_ConstructTabRenameBox(const winrt::hstring& tabText)

```

Comment:
Would we be able to do all of this in a xaml file?

Reply:
We probably _could_, but we'll leave that as an exercise for the reader? 😋